### ISSUE TITLE
Error while freeing CUDA.CuPtr

### ISSUE THREAD

**Issue by clintonTE** [JuliaGPU/CUDA.jl#254]
**Describe the bug**

I seem to be getting this error/warning about 1 out of 3 times that I run the code. The location in my code where the error appears seems arbitrary. From the error message, I tried to manually invoke gc, but that didn't help with reliably reproducing. I am open to suggestions as to how to better isolate this.

```Julia
WARNING: Error while freeing CUDA.CuPtr{Nothing}(0x0000001305604400):
ErrorException("task switch not allowed from inside gc finalizer")

Stacktrace:
 [1] try_yieldto(::typeof(Base.ensure_rescheduled), ::Base.RefValue{Task}) at .\task.jl:645
 [2] wait at .\task.jl:710 [inlined]
 [3] wait(::Base.GenericCondition{Base.Threads.SpinLock}) at .\condition.jl:106
 [4] lock(::ReentrantLock) at .\lock.jl:80
 [5] lock at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory.jl:17 [inlined]
 [6] macro expansion at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory.jl:34 [inlined]
 [7] free(::CUDA.CuPtr{Nothing}) at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory\binned.jl:406
 [8] macro expansion at C:\Users\Clinton\.julia\packages\TimerOutputs\dVnaw\src\TimerOutput.jl:206 [inlined]
 [9] macro expansion at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory.jl:274 [inlined]
 [10] macro expansion at .\util.jl:234 [inlined]
 [11] free at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory.jl:273 [inlined]
 [12] _unsafe_free!(::CUDA.CuArray{Float32,2,Nothing}) at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\array.jl:53
 [13] unsafe_free!(::CUDA.CuArray{Float32,2,Nothing}) at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\array.jl:42
 [14] gc at .\gcutils.jl:68 [inlined]
 [15] scan() at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory\binned.jl:105
 [16] macro expansion at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory\binned.jl:366 [inlined]
 [17] macro expansion at C:\Users\Clinton\.julia\packages\TimerOutputs\dVnaw\src\TimerOutput.jl:206 [inlined]
 [18] macro expansion at C:\Users\Clinton\.julia\packages\CUDA\nb3KY\src\memory\binned.jl:365 [inlined]
 [19] (::CUDA.BinnedPool.var"#4#5")() at .\task.jl:358
```

**To reproduce**

The Minimal Working Example (MWE) for this bug:
I'd like to put one together, but I am having a lot of difficulty figuring out where this is coming from in my code due to its intermittent and arbitrary appearance. 

<details><summary>Manifest.toml</summary>
<p>

```
  [6e4b80f9] BenchmarkTools v0.5.0
  [a134a8b2] BlackBoxOptim v0.5.0 #master (https://github.com/robertfeldt/BlackBoxOptim.jl.git)
  [336ed68f] CSV v0.7.0
  **[052768ef] CUDA v1.0.2 #master (https://github.com/JuliaGPU/CUDA.jl.git)**
  [6b39b394] CodecZstd v0.7.0
  [a93c6f00] DataFrames v0.21.3
  [31c24e10] Distributions v0.23.4
  [ced4e74d] DistributionsAD v0.6.1
  [2167859e] Finometrics v0.1.0 #master (https://github.com/clintonTE/Finometrics)
  [9d5cd8c9] FixedEffectModels v0.8.4
  [f6369f11] ForwardDiff v0.10.10 #master (https://github.com/JuliaDiff/ForwardDiff.jl.git)
  [38e38edf] GLM v1.3.9
  [2e9cd046] Gurobi v0.8.1
  [4076af6c] JuMP v0.21.3
  [e5e0dc1b] Juno v0.8.2
  [0fc2ff8b] LeastSquaresOptim v0.7.5 #master (https://github.com/matthieugomez/LeastSquaresOptim.jl.git)
  [c7f686f2] MCMCChains v4.0.0
  [d96e819e] Parameters v0.12.1
  [295af30f] Revise v2.7.3
  [276daf66] SpecialFunctions v0.10.3
  [2913bbd2] StatsBase v0.33.0
  [28d57a85] Transducers v0.4.36
  [3a884ed6] UnPack v1.0.1
  [112f6efa] VegaLite v2.2.0
  [fdbf4ff8] XLSX v0.7.2
  [e88e6eb3] Zygote v0.5.1 #master (https://github.com/FluxML/Zygote.jl.git)
  [ade2ca70] Dates
  [8ba89e20] Distributed
  [b77e0a4c] InteractiveUtils
  [37e2e46d] LinearAlgebra
  [44cfe95a] Pkg
  [9a3f8284] Random
  [9e88b42a] Serialization
  [2f01184e] SparseArrays
  [10745b16] Statistics
```

</p>
</details>

**Version info**

```
Julia Version 1.4.2
Commit 44fa15b150* (2020-05-23 18:35 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
Environment:
  JULIA_EDITOR = "C:\Users\Clinton\AppData\Local\atom\app-1.48.0\atom.exe"  -a
  JULIA_NUM_THREADS = 7
```

Details on CUDA:

```
julia> CUDA.versioninfo()
CUDA toolkit 10.2.89, artifact installation
CUDA driver 11.0.0
NVIDIA driver 445.87.0

Libraries: 
- CUBLAS: 10.2.2
- CURAND: 10.1.2
- CUFFT: 10.1.2
- CUSOLVER: 10.3.0
- CUSPARSE: 10.3.1
- CUPTI: 12.0.0
- NVML: 11.0.0+445.87
- CUDNN: 7.6.5 (for CUDA 10.2.0)
- CUTENSOR: missing

Toolchain:
- Julia: 1.4.2
- LLVM: 8.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75

1 device(s):
- GeForce RTX 2070 with Max-Q Design (5.451 GiB, sm_75)
```

**Additional context**
Happy to try other debugging methods to help create a reproducer.

**Comment by maleadt** [comment]
I take it you are using multiple threads?

**Comment by maleadt** [comment]
If so, could you check if https://github.com/JuliaGPU/CUDA.jl/pull/25 helps?

**Comment by clintonTE** [comment]
Interesting- I had avoided multithreading the CPU with code using device arrays, but your comment made me double-check how some of my methods were dispatching. Turns out I was accidentally dispatching to a multi-threaded method intended for the CPU version. Fixing the dispatching got rid of the error- thanks!

To test #25 I intentionally dispatched to the threaded loop in question. The PR seemed to fix the problem- except then I couldn't reproduce the issue on Master anymore. To make matters even more annoying, there is some spooky action at a distance where the warning appears in the first pass or two AFTER installing the PR or switching to master., even in a fresh Julia session. This goes away permanently after a pass or two. Also, a few attempts by me to produce an MWE also failed to generate the warning in any version. ¯\_(ツ)_/¯

I'm going to close this issue since I can't seem to repro it at all on Master anymore (just in 1.02). Happy to check again though on another branch if helpful.
